contracts/VotingMachines/GenesisProtocol.sol

```diff
@@ -175,42 +180,36 @@ contract GenesisProtocol is IntVoteInterface,UniversalScheme {
      *              false - otherwise.
      */
     function stake(bytes32 _proposalId, uint _vote, uint _amount) external returns(bool) {
-        // 0 is not a valid vote.
-        require(_vote <= NUM_OF_CHOICES && _vote > 0);
-        require(_amount > 0);
-        if (execute(_proposalId)) {
-            return true;
-        }
-
-        Proposal storage proposal = proposals[_proposalId];
-
-        if (proposal.state != ProposalState.PreBoosted) {
-            return false;
-        }
-
-        // enable to increase stake only on the previous stake vote
-        Staker storage staker = proposal.stakers[msg.sender];
-        if ((staker.amount > 0) && (staker.vote != _vote)) {
-            return false;
-        }
-
-        uint amount = _amount;
-        Parameters memory params = parameters[proposal.paramsHash];
-        require(amount >= params.minimumStakingFee);
-        require(stakingToken.transferFrom(msg.sender, address(this), amount));
-        proposal.totalStakes[1] = proposal.totalStakes[1].add(amount); //update totalRedeemableStakes
-        staker.amount += amount;
-        staker.amountForBounty = staker.amount;
-        staker.vote = _vote;
+        return _stake(_proposalId,_vote,_amount,msg.sender);
+    }
 
-        proposal.votersStakes += (params.stakerFeeRatioForVoters * amount)/100;
-        proposal.stakes[_vote] = amount.add(proposal.stakes[_vote]);
-        amount = amount - ((params.stakerFeeRatioForVoters*amount)/100);
-        proposal.totalStakes[0] = amount.add(proposal.totalStakes[0]);
-      // Event:
-        emit Stake(_proposalId, proposal.avatar, msg.sender, _vote, _amount);
-      // execute the proposal if this vote was decisive:
-        return execute(_proposalId);
+    /**
+     * @dev staking function
+     * @param _proposalId id of the proposal
+     * @param _vote  NO(2) or YES(1).
+     * @param _amount the betting amount
+     * @param _nonce nonce value ,it is part of the signature to ensure that
+              a signature can be received only once.
+     * @param _signature  - signed data by the staker
+     * @return bool true - the proposal has been executed
+     *              false - otherwise.
+     */
+    function stakeWithSignature(
+        bytes32 _proposalId,
+        uint _vote,
+        uint _amount,
+        uint _nonce,
+        bytes _signature
+        )
+        external
+        returns(bool)
+        {
+        require(stakeSignatures[_signature] == false);
+        bytes32 prefixedHash = prefixed(keccak256(address(this),_proposalId,_vote,_amount,_nonce));
+        address staker = prefixedHash.recover(_signature);
```

Comment:
If the signature is invalid the address will be garbage, maybe we should require `staker` to be equal to `msg.sender` or `tx.origin` to prevent mistakes.

Reply:
if the signature is invalid the return address is 0.

Reply:
I think `ecrecover` may still return garbage if `r` or `s` are invalid.

Reply:
if the address is garbage the tx will revert .No approval and no funds..

Reply:
comment added a936d94